furnace.c

```diff
@@ -609,6 +695,76 @@ do_stdio_work(furnace_context_t* ctx, bool deadline_met)
   }
 }
 
+#if CONFIG_AUTO == CONFIG_AUTO_MAPPER
+
+static void
+do_mapper_work(furnace_context_t *ctx)
+{
+  const bool deadline_met = get_absolute_time() > ctx->mapper.deadline;
+  if(ctx->mapper.is_enabled){
+    if(ctx->cur_temp >= MAX_TEMP){
+      const char msg[] = "cur_temp has reached MAX_TEMP, enabling auto and steering temp towards FALLBACK_TEMP!\r\n";
+      const size_t msg_len = sizeof(msg)-1;
+      tcp_server_send_data(
+        ctx,
+        ctx->tcp.client_pcb,
+        (uint8_t*)msg,
+        msg_len
+      );
+
+      ctx->mapper.is_enabled = false;
+      ctx->pilot.is_enabled = true;
+      ctx->pilot.des_temp = FALLBACK_TEMP;
+    } else if (deadline_met){
+      if(ctx->cur_temp <= ctx->mapper.max_pwm_temp){
+        char buffer[MAPPER_STATUS_SIZE];
+
+        const int size = format_mapper(buffer, ctx);
+        tcp_server_send_data(
+          ctx,
+          ctx->tcp.client_pcb,
+          (uint8_t*)buffer,
+          size
+        );
+
+        const unsigned pwm = ctx->pwm_level + 1;
+
+        const int res = set_pwm_safe(FURNACE_FIRE_PIN, ctx, pwm);
+
+        if( res == -1 ) {
+          const char msg[] = "set_pwm_safe: unexpected pin argument!\r\n";
+          const size_t msg_len = sizeof(msg)-1;
+          tcp_server_send_data(
+            ctx,
+            ctx->tcp.client_pcb,
+            (uint8_t*)buffer,
+            size
+          );
+        } else if ( res == 1 ) {
+          const char msg[] = "pwm_level has reached MAX_PWM, enabling auto and steering temp towards FALLBACK_TEMP!\r\n";
+          const size_t msg_len = sizeof(msg)-1;
+          tcp_server_send_data(
+            ctx,
+            ctx->tcp.client_pcb,
+            (uint8_t*)buffer,
+            size
+          );
+
+          ctx->mapper.is_enabled = false;
+          ctx->pilot.is_enabled = true;
+          ctx->pilot.des_temp = FALLBACK_TEMP;
+        }
+      } else {
+        ctx->mapper.max_pwm_temp = ctx->cur_temp;
+      }
```

Comment:
That's some complex logic. Can we move this to a separate function like `mapper_deadline_()` and here just do:
```C
    if(ctx->cur_temp >= MAX_TEMP){
      const char msg[] = "cur_temp has reached MAX_TEMP, enabling auto and steering temp towards FALLBACK_TEMP!\r\n";
      const size_t msg_len = sizeof(msg)-1;
      tcp_server_send_data(
        ctx,
        ctx->tcp.client_pcb,
        (uint8_t*)msg,
        msg_len
      );

      ctx->mapper.is_enabled = false;
      ctx->pilot.is_enabled = true;
      ctx->pilot.des_temp = FALLBACK_TEMP;
      
     return; /* Return to simplify the logic and get rid of chain of if-else statements. */
    }

    if (deadline_met)
        mapper_deadline_();
```

We probably want to move the `ctx->cur_temp >= MAX_TEMP` logic to a separate function too and just call it here to get rid of another level of indentation.

```C
static void
mapper_maxtemp_reached_(ctx)
{
    /* ... */
}

static void
mapper_deadline_(ctx)
{
    /* ... */
}

static void
do_mapper_work(furnace_context_t *ctx)
{
    if (!ctx->mapper_enabled)
        return;

    if (ctx->cur_temp >= MAX_TEMP)
        return mapper_maxtemp_reached_();

    const bool deadline_met = /* ... */;
    if (deadline_met)
        return mapper_deadline_();
}
```

Notice the `_` at the end of the helper functions, that indicate that the function may be running with some assumptions and user should make sure they know what they are doing.